### PR TITLE
[TTS] FastPitch speaker encoder

### PR DESCRIPTION
**Pull request by hsiehjackson**
# What does this PR do ?

Adds FastPitch speaker encoder with reference audio and reference speaker embedding

**Collection**: TTS

# Changelog 
- Adds multi-speaker FastPitch pre-training with 1) Speaker verification model speaker embedding (reference speaker embedding) and 2) Global Style Tokens (reference audio)
    - `nemo/collections/tts/data/tts_dataset.py`
    - `nemo/collections/tts/torch/tts_data_types.py`
    - `nemo/collections/tts/models/fastpitch.py`
    - `nemo/collections/tts/modules/fastpitch.py`
    - `nemo/collections/tts/modules/submodules.py`
- Modified config and fine-tuning python script
    - `examples/tts/conf/fastpitch_speaker_adaptation.yaml`
    - `examples/tts/fastpitch_finetune_adapters.py`
 
# Usage

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/tts/modules/submodules.py [hunk 1]
```diff
@@ -410,3 +418,254 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
...
+        for layer in self.layers:
+            x = x * x_masks
+            x = layer(x)
+            x_lens = self.calculate_post_conv_lengths(x_lens)
+            x_masks = self.lengths_to_masks(x_lens).unsqueeze(2).unsqueeze(3)
+
+        x = x * x_masks
+        x = x.contiguous().view(x.shape[0], x.shape[1], -1)
+        self.gru.flatten_parameters()
+        _, x = self.gru(x)
```
Comment (rlangman) [inline nemo/collections/tts/modules/submodules.py:548]: RNN input needs to be padded_sequence (https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.pack_padded_sequence.html). Simply masking X does not work because the GRU will still receive the 0 entries as inputs.

File: nemo/collections/tts/modules/submodules.py [hunk 2]
```diff
@@ -410,3 +418,254 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
...
+
+    def forward(self, inputs, inputs_lengths):
+        # BMW -> BWMC (M: mels)
+        x = inputs.transpose(1, 2).unsqueeze(3)
+        x_lens = inputs_lengths
+        x_masks = self.lengths_to_masks(x_lens).unsqueeze(2).unsqueeze(3)
+
+        for layer in self.layers:
+            x = x * x_masks
+            x = layer(x)
```
Comment (rlangman) [inline nemo/collections/tts/modules/submodules.py:541]: Nitpick: slightly cleaner to have Conv2DReLUNorm take mask as input argument and mask internally (which reminds future users of the Conv2DReLUNorm that masking is required).

File: nemo/collections/tts/modules/submodules.py [hunk 3]
```diff
@@ -410,3 +418,254 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
...
+    def output_types(self):
+        return {
+            "style_emb": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inputs):
+        bs = inputs.size(0)
+        query = inputs.unsqueeze(1)
+        key = F.tanh(self.tokens).unsqueeze(0).expand(bs, -1, -1)
+        value = F.tanh(self.tokens).unsqueeze(0).expand(bs, -1, -1)
```
Comment (rlangman) [inline nemo/collections/tts/modules/submodules.py:623]: Nitpick: only need to compute tanh(self.tokens) once.

File: nemo/collections/tts/modules/submodules.py [hunk 4]
```diff
@@ -410,3 +418,244 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
...
+    def forward(self, spk_emb, reference_spec=None, reference_spec_lens=None):
+        x = spk_emb
+
+        # Project lookup speaker embedding
+        if self.lookup_emb_projection_module is not None:
+            out = self.lookup_emb_projection_module(spk_emb)
+            x = out if x is None else x + out
+
+        # Get GST based speaker embedding
+        if self.gst_module is not None and reference_spec is not None and reference_spec_lens is not None:
```
Comment (rlangman) [inline nemo/collections/tts/modules/submodules.py:657]: Should throw error if gst_module is not None and reference_spec is not provided.

This is also where you can include the GST errors/warnings that are in the FastPitchModel class right now.

File: nemo/collections/tts/data/dataset.py [hunk 5]
```diff
@@ -522,6 +529,16 @@ def _pad_wav_to_multiple(self, wav):
                 )
         return wav
 
+    # Random sample a reference index from the same speaker
+    def sample_reference_index(self, curr_index, n=1):
+        speaker_id = self.data[curr_index]["speaker_id"]
+        if len(self.speaker_to_index_map[speaker_id]) > 1:
+            reference_pool = self.speaker_to_index_map[speaker_id] - set([curr_index])
+        else:
+            reference_pool = self.speaker_to_index_map[speaker_id]
```
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/data/dataset.py]: Yeap. It would be fine if we have enough utterances per speaker. Fixed.

File: nemo/collections/tts/data/dataset.py [hunk 6]
```diff
@@ -683,6 +700,20 @@ def __getitem__(self, index):
         if SpeakerID in self.sup_data_types_set:
             speaker_id = torch.tensor(sample["speaker_id"]).long()
 
+        reference_audio, reference_audio_length = None, None
+        if ReferenceAudio in self.sup_data_types_set:
+            if self.speaker_to_index_map is not None:
+                reference_index = self.sample_reference_index(index, n=1)[0]
```
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/data/dataset.py]: I create `self.speaker_to_index_map` when we set `ReferenceAudio`, so I remove the condition.

File: nemo/collections/tts/modules/submodules.py (hunk 1 above)
Comment (hsiehjackson) [reply nemo/collections/tts/modules/submodules.py]: That makes sense! Fixed.

File: nemo/collections/tts/modules/submodules.py (hunk 4 above)
Comment (hsiehjackson) [reply nemo/collections/tts/modules/submodules.py:662]: Not sure is it good if we raise error inside the module? Since if we don't have that, the module will directly do. I write in FastPitchModel class to have an early check.

File: nemo/collections/tts/losses/aligner_loss.py [hunk 7]
```diff
@@ -58,9 +58,7 @@ def forward(self, attn_logprob, in_lens, out_lens):
         # Convert to log probabilities
         # Note: Mask out probs beyond key_len
         key_inds = torch.arange(max_key_len + 1, device=attn_logprob.device, dtype=torch.long)
-        attn_logprob.masked_fill_(
-            key_inds.view(1, 1, -1) > key_lens.view(1, -1, 1), -float("inf")  # key_inds >= key_lens+1
-        )
+        attn_logprob.masked_fill_(key_inds.view(1, 1, -1) > key_lens.view(1, -1, 1), -1e15)  # key_inds >= key_lens+1
```
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/losses/aligner_loss.py:61]: Yes. I found normal FastPitch training will get loss to `nan` if I use -inf here.

File: nemo/collections/tts/modules/submodules.py [hunk 8]
```diff
@@ -410,3 +418,254 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
                 output = output + res_skip_acts
 
         return self.end(output)
+
+
+class ConditionalLayerNorm(torch.nn.Module):
+    def __init__(self, normalized_shape, spk_emb_dim, eps=1e-5, bias=True, condition=False):
```
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/modules/submodules.py:425]: Changed in this [PR](https://github.com/NVIDIA/NeMo/pull/6416)

File: nemo/collections/tts/losses/aligner_loss.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/losses/aligner_loss.py:61]: Restore back to (-inf). This problem is not happened again.

File: nemo/collections/tts/modules/transformer.py [hunk 9]
```diff
@@ -68,17 +70,17 @@ def __init__(self, d_model, d_inner, kernel_size, dropout, pre_lnorm=False):
             nn.Conv1d(d_inner, d_model, kernel_size[1], 1, (kernel_size[1] // 2)),
             nn.Dropout(dropout),
         )
-        self.layer_norm = nn.LayerNorm(d_model)
+        self.layer_norm = ConditionalLayerNorm(d_model, spk_emb_dim=d_model, condition=condition_lnorm)
```
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/modules/transformer.py:73]: discuss in this [PR](https://github.com/NVIDIA/NeMo/pull/6416)

File: nemo/collections/tts/modules/submodules.py [hunk 10]
```diff
@@ -410,3 +418,254 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
...
+        super(ConditionalLayerNorm, self).__init__()
+        self.normalized_shape = normalized_shape
+        self.eps = eps
+        self.spk_emb_dim = spk_emb_dim
+        self.condition = condition
+
+        if condition:
```
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/modules/submodules.py:432]: Discuss in this [PR](https://github.com/NVIDIA/NeMo/pull/6416)

File: nemo/collections/tts/modules/submodules.py (hunk 4 above)
Comment (rlangman) [reply nemo/collections/tts/modules/submodules.py:662]: From an abstraction standpoint, nothing outside of the module should require any knowledge of what happens inside the model. Having error checks inside is also pretty foolproof.

File: nemo/collections/tts/models/fastpitch.py [hunk 11]
```diff
@@ -141,16 +143,31 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         speaker_emb_condition_prosody = cfg.get("speaker_emb_condition_prosody", False)
         speaker_emb_condition_decoder = cfg.get("speaker_emb_condition_decoder", False)
         speaker_emb_condition_aligner = cfg.get("speaker_emb_condition_aligner", False)
+        speaker_encoder = instantiate(self._cfg.get("speaker_encoder", None))
         energy_embedding_kernel_size = cfg.get("energy_embedding_kernel_size", 0)
         energy_predictor = instantiate(self._cfg.get("energy_predictor", None))
 
+        """Check use both speaker encoder and reference data"""
+        if self._train_dl is not None:
```
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/models/fastpitch.py]: Check logic move to module itself

File: nemo/collections/tts/modules/submodules.py [hunk 12]
```diff
@@ -507,3 +514,240 @@ def forward(self, inputs, conditioning=None):
...
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "style_emb": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inputs):
+        bs = inputs.size(0)
```
Comment (rlangman) [inline nemo/collections/tts/modules/submodules.py:550]: Nitpick: use clearer name batch_size

File: nemo/collections/tts/modules/submodules.py [hunk 13]
```diff
@@ -507,3 +514,240 @@ def forward(self, inputs, conditioning=None):
...
+    def forward(self, inp, inp_lengths):
+        style_embedding = self.reference_encoder(inp, inp_lengths)
+        gst = self.style_attention(style_embedding)
+        return gst
+
+
+class LookupTable(torch.nn.Module):
+    """
+    LookupTable based Speaker Embedding
+    """
```
Comment (rlangman) [inline nemo/collections/tts/modules/submodules.py:696]: Maybe rename it to SpeakerLookupTable to make it clear how it differs from torch.nn.Embedding.

File: nemo/collections/tts/modules/submodules.py [hunk 14]
```diff
@@ -507,3 +514,240 @@ def forward(self, inputs, conditioning=None):
...
+
+        # Get Lookup table speaker embedding
+        if self.lookup_module is not None and speaker is not None:
+            embs = self.lookup_module(speaker)
+
+        # Get GST based speaker embedding
+        if self.gst_module is not None and reference_spec is not None and reference_spec_lens is not None:
+            out = self.gst_module(reference_spec, reference_spec_lens)
+            embs = out if embs is None else embs + out
+        elif self.gst_module is not None and reference_spec is None and reference_spec_lens is None:
```
Comment (rlangman) [inline nemo/collections/tts/modules/submodules.py:746]: Can simplify to:

```
if self.gst_module is not None:
    if reference_spec is None or reference_spec_lens is None:
        raise ValueError()
```

File: examples/tts/conf/fastpitch_align_44100_adapter.yaml [hunk 15]
```diff
@@ -35,10 +35,9 @@ model:
   learn_alignment: true
   bin_loss_warmup_epochs: 100
   
-  n_speakers: 1
+  n_speakers: 1 # [TODO] removed after we change previous config
```
Comment (rlangman) [inline examples/tts/conf/fastpitch_align_44100_adapter.yaml:38]: Should we make this `??? ` so user knows they have to provide it?

File: examples/tts/conf/fastpitch_align_44100_adapter.yaml [hunk 16]
```diff
@@ -248,6 +247,28 @@ model:
     n_layers: 2
     condition_types: [ "add", "layernorm" ] # options: [ "add", "cat", "layernorm" ]
 
+  speaker_encoder:
+    _target_: nemo.collections.tts.modules.submodules.SpeakerEncoder
+    lookup_module:
+      _target_: nemo.collections.tts.modules.submodules.LookupTable
+      n_speakers: 1
```
Comment (rlangman) [inline examples/tts/conf/fastpitch_align_44100_adapter.yaml:254]: Use variable ${n_speakers}

File: examples/tts/conf/fastpitch_align_44100_adapter.yaml (hunk 15 above)
Comment (hsiehjackson) [reply examples/tts/conf/fastpitch_align_44100_adapter.yaml:38]: Yes, it can be `???`. But I think this argument can be placed into `speaker_encoder.lookup_module`.

Comment (hsiehjackson) [reply examples/tts/conf/fastpitch_align_44100_adapter.yaml:38]: I see. But this parameter is not used if we don't use lookup module. So, do we still need `???`?

Comment (rlangman) [reply examples/tts/conf/fastpitch_align_44100_adapter.yaml]: The config is just an example, but it should be a complete example. If the example includes the lookup module, then n_speakers should be ??? because the adapter system does not make sense with n_speakers = 1. If there were a different example config without lookup module, then we could remove it.

Comment (rlangman) [reply examples/tts/conf/fastpitch_align_44100_adapter.yaml]: Ok, that is fine.

**Review by rlangman — APPROVED** [review]
LGTM

**Review by subhankar-ghosh — APPROVED** [review]
